File: js/String/11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python
#
# [11] Container With Most Water
#

# @lc code=start
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        res = 0

        for l in range(0, len(height)):
            for r in range(l + 1, len(height)):
                area = (r - l) * min(height[l], height[r]) # 會選擇左右兩邊最小的原因：水位的面積高度只會到較低的那個高度
                res = max(res, area)
        return res
        
# @lc code=end
a =Solution()
a.maxArea([1,8,6,2,5,4,8,3,7])

File: js/Array/11.container-with-most-water.js
/*
 * @lc app=leetcode id=11 lang=javascript
 *
 * [11] Container With Most Water
 */

// @lc code=start
/**
 * @param {number[]} heights
 * @return {number}
 */
var maxArea = function (heights) {
    let maxArea = 0;
    p1 = 0;
    p2 = heights.length - 1;
    while (p1 < p2) {
        const height = Math.min(heights[p1], heights[p2]);
        const width = p2 - p1;
        const area = height * width;

        maxArea = Math.max(maxArea, area);
        if (heights[p1] <= heights[p2]) {
            p1++;
        } else if (heights[p1] > heights[p2]) {
            p2--;
        }
    }
    return maxArea;
};
// @lc code=end

